### PR TITLE
Basic secured tls connection for migrations

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Use self-signed certificates and a secret to secure migrations until
more sophisticated mechanisms are available. It uses the already used mechanism of storing self signed  certificates in a secret, like it is done for virt-api.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Secure migrations with self-signed certs
```

### CODE REVIEW

File: cmd/virt-handler/virt-handler.go [hunk 1]
```diff
@@ -236,6 +357,42 @@ func (app *virtHandlerApp) AddFlags() {
...
+	}
+
+	certPool := x509.NewCertPool()
+
+	for _, crt := range caCert {
+		certPool.AddCert(crt)
+	}
+
+	app.migrationTLSConfig = &tls.Config{
+		RootCAs: certPool,
```
Comment (davidvossel) [inline cmd/virt-handler/virt-handler.go]: RootCAs are used to validate server connections. 

I'm pretty sure we need to set ClientCAs as well in order to verify client connections when "RequireAndVerifyClientCert" is used.

File: cmd/virt-handler/virt-handler.go [hunk 2]
```diff
@@ -83,10 +98,106 @@ type virtHandlerApp struct {
...
+					v1.AppLabel: "virt-api-aggregator",
+				},
+			},
+			Type: "Opaque",
+			Data: map[string][]byte{
+				serverCertBytesValue:  app.serverCertBytes,
+				serverKeyBytesValue:   app.serverKeyBytes,
+				clientCertBytesValue:  app.clientCertBytes,
+				clientKeyBytesValue:   app.clientKeyBytes,
+				signingCertBytesValue: app.signingCertBytes,
```
Comment (davidvossel) [inline cmd/virt-handler/virt-handler.go]: The only reason we'd need to save "signingCertBytes" is if we plan on generating new certs, which this pr doesn't do.

File: manifests/generated/rbac-kubevirt.authorization.k8s.yaml.in [hunk 3]
```diff
@@ -285,10 +285,17 @@ rules:
...
   verbs:
   - get
+- apiGroups:
+  - ""
+  resources:
+  - secrets
+  verbs:
+  - get
+  - create
+  - update
```
Comment (davidvossel) [inline manifests/generated/rbac-kubevirt.authorization.k8s.yaml.in]: I think we just need 'get' and 'create'

also 'create' could be limited to the namespace kubevirt infra is running in.  I don't think that's a big deal though as long as we don't have permissions to 'update' in any namespace.

File: pkg/virt-handler/migration-proxy/migration-proxy.go [hunk 4]
```diff
@@ -304,12 +309,19 @@ func NewTargetProxy(tcpBindAddress string, tcpBindPort int, libvirtdSocketPath s
...
 }
 
 func (m *migrationProxy) createTcpListener() error {
-	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", m.tcpBindAddress, m.tcpBindPort))
+	var listener net.Listener
+	var err error
+	if m.tlsConfig != nil {
+		listener, err = tls.Listen("tcp", fmt.Sprintf("%s:%d", m.tcpBindAddress, m.tcpBindPort), m.tlsConfig)
+	} else {
+		listener, err = net.Listen("tcp", fmt.Sprintf("%s:%d", m.tcpBindAddress, m.tcpBindPort))
```
Comment (davidvossel) [inline pkg/virt-handler/migration-proxy/migration-proxy.go]: i think I mentioned this in the previous pr.

what would you think about only allowing a non TLS listener if bindAddr == 127.0.0.1?

File: cmd/virt-handler/virt-handler.go (hunk 1 above)
Comment (rmohr) [reply cmd/virt-handler/virt-handler.go]: Done.

File: cmd/virt-handler/virt-handler.go (hunk 2 above)
Comment (rmohr) [reply cmd/virt-handler/virt-handler.go]: We need it to verify that the peer certificate is signed by this ca.

File: manifests/generated/rbac-kubevirt.authorization.k8s.yaml.in (hunk 3 above)
Comment (rmohr) [reply manifests/generated/rbac-kubevirt.authorization.k8s.yaml.in]: Reverted the cluster-wide  permissions to its old state (*just* get), and added create for the install namespace.

File: pkg/virt-handler/migration-proxy/migration-proxy.go (hunk 4 above)
Comment (rmohr) [reply pkg/virt-handler/migration-proxy/migration-proxy.go]: Done.

File: cmd/virt-handler/virt-handler.go [hunk 5]
```diff
@@ -236,6 +331,62 @@ func (app *virtHandlerApp) AddFlags() {
...
+				return fmt.Errorf("no client certificate provided.")
+			}
+
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("failed to parse peer certificate: %v", err)
+			}
+			_, err = c.Verify(x509.VerifyOptions{
+				Roots:     certPool,
+				KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
```
Comment (davidvossel) [inline cmd/virt-handler/virt-handler.go]: ExtKeyUsageServerAuth | ExtKeyUsageClientAuth here?

File: cmd/virt-handler/virt-handler.go [hunk 6]
```diff
@@ -236,6 +331,62 @@ func (app *virtHandlerApp) AddFlags() {
...
+			})
+
+			if err != nil {
+				return fmt.Errorf("could not verify peer certificate: %v", err)
+			}
+			return nil
+		},
+		ClientAuth: tls.RequireAnyClientCert,
```
Comment (davidvossel) [inline cmd/virt-handler/virt-handler.go]: RequireAndVerifyClientCert instead?

Comment (rmohr) [reply cmd/virt-handler/virt-handler.go]: Changed. Luckily it makes no difference. In both cases the verify function is called.

File: cmd/virt-handler/virt-handler.go (hunk 5 above)
Comment (rmohr) [reply cmd/virt-handler/virt-handler.go]: Done.

File: tests/migration_test.go [hunk 7]
```diff
@@ -523,6 +524,98 @@ var _ = Describe("Migrations", func() {
...
+				handler, err := kubecli.NewVirtHandlerClient(virtClient).ForNode(vmi.Status.MigrationState.TargetNode).Pod()
+				Expect(err).ToNot(HaveOccurred())
+				// The port-forwarder tears down after each check, but may be too slow, so use different ports on fast checks
+				i := 0
+				for port, _ := range vmi.Status.MigrationState.TargetDirectMigrationNodePorts {
+					func() {
+						stopChan := make(chan struct{})
+						defer close(stopChan)
+						Expect(tests.ForwardPorts(handler, []string{fmt.Sprintf("4321%d:%d", i, port)}, stopChan, 10*time.Second)).To(Succeed())
+						_, err = tls.Dial("tcp", fmt.Sprintf("localhost:4321%d", i), nil)
```
Comment (davidvossel) [inline tests/migration_test.go]: need a functional test that validates we reject a client cert that isn't signed by the CA .

File: pkg/virt-handler/migration-proxy/migration-proxy.go (hunk 4 above)
Comment (davidvossel) [reply pkg/virt-handler/migration-proxy/migration-proxy.go]: local migrations don't happen. This 127.0.0.1 link is to get around src libvirt not being able to connect to destination libvirt via a unix socket. We have to create fake tcp endpoint and pipe it through the unix socket locally to get out of the pod... not super great :/

Comment (cynepco3hahue) [reply pkg/virt-handler/migration-proxy/migration-proxy.go]: I see, thanks for the clarification, still need to pass over migration-proxy code :worried:

File: tests/migration_test.go [hunk 8]
```diff
@@ -523,6 +526,114 @@ var _ = Describe("Migrations", func() {
...
+			var originalMigrationConfig string
+			var kubevirtConfig = "kubevirt-config"
+
+			BeforeEach(func() {
+				// update migration timeouts
+				options = metav1.GetOptions{}
+				cfgMap, err = virtClient.CoreV1().ConfigMaps(namespaceKubevirt).Get(kubevirtConfig, options)
+				Expect(err).ToNot(HaveOccurred())
+				originalMigrationConfig = cfgMap.Data["migrations"]
+				cfgMap.Data["migrations"] = `{"bandwidthPerMigration" : "1Mi"}`
```
[the remark this answers is not in the document]
Comment (rmohr) [reply tests/migration_test.go]: I don't have to. Just running stress should be enough aleady. Probably no need to throttle it even more.

**Comment by davidvossel** [comment]
We'll want to follow up with improvements involving cert and CA rotation eventually. For now, this gives us a foundation for securing live migration. I see this as an important first step.

**Pull request merged**